merkle.go

```diff
@@ -0,0 +1,189 @@
+package fastmerkle
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+)
+
+var (
+	errEmptyDataSet = errors.New("empty data set provided")
+)
+
+// GenerateMerkleTree generates a Merkle tree based on the input data
+func GenerateMerkleTree(inputData [][]byte) (*MerkleTree, error) {
+	// Check if the data set is valid
+	if !isValidDataSet(inputData) {
+		return nil, errEmptyDataSet
+	}
+
+	// Create the worker pool and put them on standby
+	workerPool := newWorkerPool(runtime.NumCPU())
+	workerPool.startWorkerPool()
+	defer workerPool.close()
+
+	// Generate the leaves of the Merkle tree
+	nodes, leafErr := generateLeaves(inputData, workerPool)
+	if leafErr != nil {
+		return nil, fmt.Errorf(
+			"unable to generate leaf nodes, %w",
+			leafErr,
+		)
+	}
+
+	// While the root is not derived, create new hashing jobs
+	// for the worker pool
+	for len(nodes) > 1 {
```

Comment:
Just curious question.

You're splitting task horizontally and you split task at the same level.
Did you compare with the way by splitting vertically? (split left and right sub-tree recursively and spawn task)

Reply:
Honestly, the initial idea was to let the algo run by generating tasks for parent nodes, as soon as child nodes were ready. The problem I ran into there was how do you organize result storage that doesn't have that big of a memory footprint. 

The program will be most concurrent on the first level, and will decrease with each level parsed. 
Do you have any suggestions how we can improve (speed up) the processing algo even more? I'm all for it if you have some clue as to how to solve these storage issues